108: consolidate tile info

davidabrahams opened the pull request:
Fixes #108. Moves all tile info to `tiles.js`.

Things done:

- Moved all information about tiles into the `tileInfo` object in `tiles.js`. This eliminates the need for large switch statements. This also makes it easier for us to change the way we treat tiles in the future, since all their info is in one place.
- The functions that previously were large switch statements have been moved to `tiles.js`, and just look up the appropriate info in `tileInfo`. They still handle improper inputs identically.
- Moved the tests for these giant switchy functions to `tiles.spec.js`

Things we should discuss:
- The tile info is dependent on knowledge about the current player. Therefore, whenever testing anything that needs tile info, we have to mock the current player, and call `computeTileInfo` to store the appropriate values in `tileInfo`. This makes the tests really long and repetitive, and we might want to think about how to shorten them.

TODO:
- [x] comment functions in `tiles.js`
- [x] (maybe?) make tests less repetitive

EDIT:

I added a `setup` and `teardown` function in `tiles.spec.js`, which you must call at the beginning and end of any test that checks tile info. This saved a lot of lines. I also added unittests for the new functions in `tiles.js`! We're almost at 100 unittests! (we're at 95)

Edit again:

we're up to 102!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/chauncy-crib/tagprobot/110)
<!-- Reviewable:end -->

davidabrahams commented on test/map.spec.js:
```diff
@@ -381,8 +385,10 @@ test('getMapTraversabilityInCells: returns correctly with CPTL=1', t => {
   const blank = tileTypes.REGULAR_FLOOR;
 
   // initialize current player as blue
```
@billmwong I'm having trouble with this test. Could you please help me with your rewiring mastery?

davidabrahams commented on src/helpers/map.js:
```diff
@@ -159,58 +72,12 @@ export function isCNTO(tileID) {
-    default:
-      throw new Error(`Unknown tileID: ${tileID}`);
-  }
+  assert(Object.prototype.hasOwnProperty.call(tileNames, tileID), `Unknown tileID: ${tileID}`);
```
switch to lodash

davidabrahams commented on src/helpers/map.js:
```diff
@@ -1,6 +1,7 @@
+import _ from 'lodash';
```
only import specific function

davidabrahams commented on src/tiles.js:
```diff
@@ -0,0 +1,51 @@
+import { getMe, amBlue, amRed } from './helpers/player';
+
+export const tileInfo = {
```
populate this when page loads

shanek21 commented:
Looks good! I'll approve once you finish the first thing in your TODO list (comments in `tile.js`

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,152 @@
+ */
+export function isTileType(tileId, name) {
+  assert(includes(keys(tileInfo), name), `Unknown tile name: ${name}`);
+  return tileInfo[name].id === tileId;
```
This should use `getId(name)` instead of `tileInfo[name].id`

davidabrahams replied:
done

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,152 @@
+export function resetTileInfo() {
+  tileInfo = undefined;
+  tileNames = undefined;
+}
```
We can't do this with rewire?

davidabrahams replied:
we might be able to. ill look into it

davidabrahams replied:
fixed now, setup and teardown are done using rewire

shanek21 commented on src/index.js:
```diff
@@ -42,6 +43,7 @@ function waitForId(fn) {
 function start() {
   setupMe();
   setupVelocity();
+  computeTileInfo();
```
Why does this need to be here?

davidabrahams replied:
see comment above

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,152 @@
+ * amRed, so the current player must be defined. We call this function once, after our player has an
+ * id.
+ */
+export function computeTileInfo() {
```
Why is `tileInfo` in a function?

davidabrahams replied:
There's a lag between when the client side javascript loads, and when our player has an id. We only compute all the info about the tiles once. If we just assigned it to the dictionary that's in this function, then we would store all the info about tiles immediately when the client side JS loads, and it wouldn't know if a `RED_GATE` is traversable, for example. The solution is to put the initialization of `tileInfo` in a function, and then call that function once our player has an `id`. Look at line 43 in `index.js`, thats when we initialize `tileInfo`

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,152 @@
+ * the given nontraversable object tile ID is a circular nontraversable object.
+ *
+ * Circular nontraversable objects include: boosts, powerups, spikes, buttons,
+ * bombs, and active portals
```
This is my fault, but could you add the `@param {number} tileID - ` and `@return` to `isCNTO` for me?

davidabrahams replied:
i've just been using @return when that's all you need to explain the function. Your docstring already explains what the function returns, so maybe its unnecessary? I added @param. LMK about @return

shanek21 replied:
If it were me, I would also add the `@return`, but it's definitely your call. I lean towards the side of redundancy on documentation, but how redundant is too redundant?

davidabrahams replied:
@billmwong @altodyte thoughts?

billmwong replied:
You know my opinion is for more concise code

davidabrahams replied:
I think what's currently there is a medium between redundancy and conciseness, so I'll just leave it :)

shanek21 commented on test/map.spec.js:
```diff
@@ -150,119 +100,130 @@ test('fillGridWithSubgrid: throws errors for invalid inputs', t => {
-  t.same(getTileTraversabilityInCells(tileTypes.SPIKE), expected);
-
+  t.same(getTileTraversabilityInCells(getId('SPIKE')), expected);
+  teardownTiles();
```
Yikes, probably also my fault: do you have to call that thing to cleanup a rewire every time?
`MapRewireAPI.__ResetDependency__('<VARIABLE>');`?
If so, you could add those in or make an issue and I could do it later?

davidabrahams replied:
done

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,146 @@
+ * @param {String} name - the name of a tile
+ * @return - the id for the input tile name
+ */
+export function getId(name) {
```
Rename to `getTileId()` to match `isTileType()`?

davidabrahams commented:
Done

On Tue, Aug 8, 2017 at 6:39 PM Shane Kelly <notifications@github.com> wrote:

> *@shanek21* requested changes on this pull request.
> ------------------------------
>
> In src/tiles.js
> <https://github.com/chauncy-crib/tagprobot/pull/110#discussion_r132073457>
> :
>
> > + * Circular nontraversable objects include: boosts, powerups, spikes, buttons,
> + * bombs, and active portals
> + *
> + * @param {number} tileID - the ID of the tile that should be checked if it is
> + * a circular nontraversable object.
> + */
> +export function isCNTO(tileID) {
> +  return Boolean(getTileProperty(tileID, 'radius'));
> +}
> +
> +
> +/*
> + * @param {String} name - the name of a tile
> + * @return - the id for the input tile name
> + */
> +export function getId(name) {
>
> Rename to getTileId() to match isTileType()?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chauncy-crib/tagprobot/pull/110#pullrequestreview-55115968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIhRRe1ki84F7RfP53GCrS7Pl7ZlwXGhks5sWQ22gaJpZM4OupmJ>
> .
>

shanek21 commented on src/tiles.js:
```diff
@@ -0,0 +1,146 @@
+    ANGLE_WALL_3: { id: 1.3, traversable: false },
+    ANGLE_WALL_4: { id: 1.4, traversable: false },
+    REGULAR_FLOOR: { id: 2, traversable: true },
+    RED_FLAG: { id: 3, traversable: amRed() },
```
All flags should now be considered fully traversable by both teams.

This change will affect `RED_FLAG` and `BLUE_FLAG`.

shanek21 commented on test/tiles.spec.js:
```diff
@@ -0,0 +1,171 @@
+  });
+  TileRewireAPI.__Rewire__('tileInfo', mockTileInfo);
+  TileRewireAPI.__Rewire__('tileNames', mockTileNames);
+}
```
You can't just mock `amRed()` and `amBlue()` and have this magically work?

shanek21 commented on test/tiles.spec.js:
```diff
@@ -0,0 +1,171 @@
+  t.end();
+});
+
+test('getTileProperty: errors given tileIds that don\'t exist', t => {
```
pls more of a complete sentence

`errors when given tileIds that don\'t exist`?
`throws errors when given tileIds that don\'t exist`?

shanek21 commented on test/tiles.spec.js:
```diff
@@ -0,0 +1,171 @@
+  t.end();
+});
+
+test('isTraversable: correctly returns true for varying inputs', t => {
```
Didn't we decided against the phrase `varying inputs` in test descriptions? (I would like to note that I still think it's useful phrasing.)

Maybe `orthogonal inputs` sounds less lazy?

shanek21 commented on test/tiles.spec.js:
```diff
@@ -0,0 +1,171 @@
+  t.throws(() => { isTileType(1, 1); });
+  teardownTiles();
+  t.end();
+});
```
If I had my way with :white_large_square: :space_invader:, then there would be a new line before every `t.end();` and two new lines after the imports and between each of the `test( ... { });` blocks. Just saying :smile: .

shanek21 commented:
Reviewed 1 of 5 files at r1, 1 of 6 files at r2, 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/tiles.js, line 113 at r6](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g7-r6-113:-Kr8jFnPtVQFNdHZXblW:bcbqss6) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/f4945ba6323adf6ebf955636e018f40a9b0143c3/src/tiles.js#L113)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Rename to `getTileId()` to match `isTileType()`?
</blockquote></details>

:dango:

---

*[test/map.spec.js, line 11 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jaNRqGcIQgaaLVv1:-Kr8jaNRqGcIQgaaLVv2:bv2t5km) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L11)):*
> ```JavaScript
> import { setupTiles, teardownTiles } from './tiles.spec';
> import { getTileId } from '../src/tiles';
> ```

Two new lines between large scopes

---

*[test/map.spec.js, line 114 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jfGeB_PO7mGNTMQn:-Kr8jfGeB_PO7mGNTMQo:b-raoqm7) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L114)):*
> ```JavaScript
> 
>   teardownTiles();
>   t.end();
> ```

New line before `t.end();`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 7 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[test/tiles.spec.js, line 62 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-62:-Kr9RO0ZahHFSCZwV0HC:bn5fxzd) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L62)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

You can't just mock `amRed()` and `amBlue()` and have this magically work?
</blockquote></details>

You can, but then you have to call `computeTileInfo` in every test, and actually store the `tileInfo` object. Then, in the spirit of making tests not inter-dependent, you would also want to force `tileInfo` back to null at the end of every test. Additionally, per advice from @billmwong, tests should be completely independent, so we don't want every single one of our tests to rely on the fact `computeTileInfo` does what we think it should do. If it at some point starts doing something different, then all these tests will still pass. That being said, I don't think I unittested `computeTileInfo`. This is a problem. I will mark myself as blocked on this thread, to make sure I go back and do this.

Side note: is there a convenient way for PR authors to note things they will do in the future? I can't edit the GitHub top comment from within reviewable, and I don't want to have to go into GitHub at any point during the review process. Let me know if you figure this out.

@billmwong @altodyte halp on this^

---

*[test/tiles.spec.js, line 88 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-88:-Kr9SiVGeaTmInzup1Qo:bwbn6l6) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L88)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Didn't we decided against the phrase `varying inputs` in test descriptions? (I would like to note that I still think it's useful phrasing.)

Maybe `orthogonal inputs` sounds less lazy?
</blockquote></details>

Yes.....did I mess up when rebasing and I renamed a test back to it's old name? Or should I just update it now that I moved it. Either way, this is something I need to address.

---

*[test/tiles.spec.js, line 171 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-171:-Kr9TIiZI1n8af4d70hu:b-m8s6zj) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L171)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

If I had my way with :white_large_square: :space_invader:, then there would be a new line before every `t.end();` and two new lines after the imports and between each of the `test( ... { });` blocks. Just saying :smile: .
</blockquote></details>

Sorry...what do those emojis mean? Either way, will address.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 5 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/tiles.js, line 25 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g7-r7-25:-Kr9rHzkzoqjwiqjNS-6:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/src/tiles.js#L25)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

All flags should now be considered fully traversable by both teams.

This change will affect `RED_FLAG` and `BLUE_FLAG`.
</blockquote></details>

Done.

---

*[test/map.spec.js, line 11 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jaNRqGcIQgaaLVv1:-Kr9rkVOE3fsuPiU8nGk:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L11)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Two new lines between large scopes
</blockquote></details>

Done.

---

*[test/map.spec.js, line 114 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jfGeB_PO7mGNTMQn:-Kr9sf8nXWADltCQLtII:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L114)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

New line before `t.end();`
</blockquote></details>

Done.

---

*[test/tiles.spec.js, line 78 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-78:-Kr9szrB_89jQRW7p7qY:bplgi38) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L78)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

pls more of a complete sentence

`errors when given tileIds that don\'t exist`?
`throws errors when given tileIds that don\'t exist`?
</blockquote></details>

Done. isn't "errors" a verb? haha

---

*[test/tiles.spec.js, line 88 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-88:-Kr9u8SJ7vgJhlBf-_h-:b-4vk8mv) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L88)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Yes.....did I mess up when rebasing and I renamed a test back to it's old name? Or should I just update it now that I moved it. Either way, this is something I need to address.
</blockquote></details>

Given that #111 exists, I'm going to call this resolved, since there's no need for this PR to address multiple inputs.

---

*[test/tiles.spec.js, line 171 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-171:-Kr9uMu7vf6ZOySC7pZR:b-mchh76) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L171)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Sorry...what do those emojis mean? Either way, will address.
</blockquote></details>

Done! I gotchu man. There's whitespace everywhere now. you owe me a :dango: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 5 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/tiles.js, line 113 at r6](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g7-r6-113:-Kr9wtPyMRC3oMAbjkHl:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/f4945ba6323adf6ebf955636e018f40a9b0143c3/src/tiles.js#L113)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

:dango:
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

shanek21 commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[test/map.spec.js, line 11 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jaNRqGcIQgaaLVv1:-KrAszF3too8PWIHATwq:b-k4k55j) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L11)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thank you for humoring me :)

---

*[test/tiles.spec.js, line 62 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-62:-KrAtrRyqtJdMYHon4Ee:b-orfzsp) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L62)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

You can, but then you have to call `computeTileInfo` in every test, and actually store the `tileInfo` object. Then, in the spirit of making tests not inter-dependent, you would also want to force `tileInfo` back to null at the end of every test. Additionally, per advice from @billmwong, tests should be completely independent, so we don't want every single one of our tests to rely on the fact `computeTileInfo` does what we think it should do. If it at some point starts doing something different, then all these tests will still pass. That being said, I don't think I unittested `computeTileInfo`. This is a problem. I will mark myself as blocked on this thread, to make sure I go back and do this.

Side note: is there a convenient way for PR authors to note things they will do in the future? I can't edit the GitHub top comment from within reviewable, and I don't want to have to go into GitHub at any point during the review process. Let me know if you figure this out.

@billmwong @altodyte halp on this^
</blockquote></details>

Does making a blocking comment not work?

---

*[test/tiles.spec.js, line 78 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-78:-KrAu02nYUlu3pUgPjIJ:bk7xyge) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L78)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Done. isn't "errors" a verb? haha
</blockquote></details>

In the right context, yes. Here, to me at least, it reads like a noun.

---

*[test/tiles.spec.js, line 88 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-88:-KrAu_AED-bTYScaj8pJ:b-2aotnu) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L88)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Given that #111 exists, I'm going to call this resolved, since there's no need for this PR to address multiple inputs.
</blockquote></details>

OK that seems fine, though it seems to me that there is a difference between fixing an issue and not making an issue worse.

---

*[test/tiles.spec.js, line 171 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-171:-KrAunIm888XAITqmuRa:b-72j55e) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L171)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Done! I gotchu man. There's whitespace everywhere now. you owe me a :dango: 
</blockquote></details>

OK
:dango: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[test/tiles.spec.js, line 62 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-62:-KrBRy6krNKs6GbPI9Ek:b-ws24iw) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L62)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Does making a blocking comment not work?
</blockquote></details>

It does, but I thought then that shows up for you as something you need to address?

---

*[test/tiles.spec.js, line 88 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-88:-KrBS96miNhq_WLqoGAz:b7ce97b) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L88)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

OK that seems fine, though it seems to me that there is a difference between fixing an issue and not making an issue worse.
</blockquote></details>

This test was moved from from another file, so it already existed. To me, this is just equally bad as it was before... haha

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[test/map.spec.js, line 11 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jaNRqGcIQgaaLVv1:-KrBSXUvI99836OaGtNs:bgs4bb3) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L11)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Thank you for humoring me :)
</blockquote></details>

You're still "not ok" :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

shanek21 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/tiles.spec.js, line 62 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-62:-KrBbUytKHJyNXfnLvKY:bdyzser) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L62)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

It does, but I thought then that shows up for you as something you need to address?
</blockquote></details>

Something like that, I'm still not exactly sure hahaha

---

*[test/tiles.spec.js, line 88 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-88:-KrBbaKXFNxewANSsdvZ:b-d3oorx) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L88)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

This test was moved from from another file, so it already existed. To me, this is just equally bad as it was before... haha
</blockquote></details>

Ah, makes sense

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrCGokbjzb9SBf67Icr:-KrCGnXrUcJtQV8KbJUX:bsqysa5) (no related file):*
TODO:

- [ ] unittest `computeTileInfo`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrCGokbjzb9SBf67Icr:-KrJUo4GyuQ3EGIBriNR:b-7rl0jq) (no related file):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

TODO:

- [ ] unittest `computeTileInfo`
</blockquote></details>

done!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
Reviewed 1 of 5 files at r1, 1 of 6 files at r2.
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/tiles.js, line 110 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJjENdKZFynhp2j3H_:-KrJjENeYtJH6l4vunTj:box4qtc) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L110)):*
> ```JavaScript
>  * a circular nontraversable object.
>  */
> export function isCNTO(tileID) {
> ```

This might be just me, but all these functions seem a bit overkill... Why can't we just export the tile info object and then directly access it ?

---

*[src/tiles.js, line 131 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJiJ7Ur_EW42H4D0og:-KrJiJ7Ur_EW42H4D0oh:b-yeepuo) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L131)):*
> ```JavaScript
>  */
> export function isTileType(tileId, name) {
>   assert(includes(keys(tileInfo), name), `Unknown tile name: ${name}`);
> ```

Get tile id already checks for this 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/tiles.js, line 110 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJjENdKZFynhp2j3H_:-KrJmeaxbHaZv-7E5J1_:b-14zsm1) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L110)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

This might be just me, but all these functions seem a bit overkill... Why can't we just export the tile info object and then directly access it ?
</blockquote></details>

I put the functions here because all the code shane wrote relied on them, so i wanted to keep as few diffs in the PR as possible.

One nice thing about writing the wrapper functions is we can be very explicit about asserts and throw errors if people do things they shouldnt, but I could go either way on this.

Your thoughts?

Which should i do

- leave as is
- open issue
-fix in this PR

---

*[src/tiles.js, line 131 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJiJ7Ur_EW42H4D0og:-KrJnCmLRjR2mMfvLtYc:b-z8zx9w) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L131)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

Get tile id already checks for this 
</blockquote></details>

you're right. idk why i did this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/tiles.js, line 131 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJiJ7Ur_EW42H4D0og:-KrK7HBsmacChNbUyPkH:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L131)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

you're right. idk why i did this
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

shanek21 commented:
Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/tiles.js, line 110 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJjENdKZFynhp2j3H_:-KrLdSpq8gxXvOgLgMFv:bwwocyv) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L110)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

I put the functions here because all the code shane wrote relied on them, so i wanted to keep as few diffs in the PR as possible.

One nice thing about writing the wrapper functions is we can be very explicit about asserts and throw errors if people do things they shouldnt, but I could go either way on this.

Your thoughts?

Which should i do

- leave as is
- open issue
-fix in this PR
</blockquote></details>

I like the wrapper functions, but would be fine with just directly accessing the object.

---

*[test/tiles.spec.js, line 95 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrLedTwmAe-cyRtsi3z:-KrLedTwmAe-cyRtsi4-:bazbjjo) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/test/tiles.spec.js#L95)):*
> ```JavaScript
>   t.true(mockAmRed.calledOnce);
>   t.true(mockAmBlue.calledOnce);
>   t.end();
> ```

I can't tell if there is a new line before `t.end();` here. If not, add pls

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/tiles.js, line 110 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJjENdKZFynhp2j3H_:-KrMntJv4tYPRPzyZ3dE:bcemvp1) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L110)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

I like the wrapper functions, but would be fine with just directly accessing the object.
</blockquote></details>

I think I ended up doing something that's in the happy medium. You still don't directly access the object, but instead of calling, say, `isTraversable`, you do `propertyFromId(tileID, property)`. You can also do `propertyFromName(tileName, property)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[test/tiles.spec.js, line 95 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrLedTwmAe-cyRtsi3z:-KrMoTCQc1maGWXIP_ar:bd510ko) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/test/tiles.spec.js#L95)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

I can't tell if there is a new line before `t.end();` here. If not, add pls
</blockquote></details>

There is one. No idea why it's not showing up? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrCGokbjzb9SBf67Icr:-KrT4xnvZoY0YNTd_4Ip:bcbqss6) (no related file):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

done!
</blockquote></details>

:dango:

---

*[src/tiles.js, line 110 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJjENdKZFynhp2j3H_:-KrT52RP7Sjw1eZQaU2x:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L110)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

I think I ended up doing something that's in the happy medium. You still don't directly access the object, but instead of calling, say, `isTraversable`, you do `propertyFromId(tileID, property)`. You can also do `propertyFromName(tileName, property)`
</blockquote></details>

Done.

---

*[src/tiles.js, line 131 at r10](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrJiJ7Ur_EW42H4D0og:-KrT52-ELOzh6c8WEcoY:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/c6c96a53229e517f5d61f36f198522df1ae9ca72/src/tiles.js#L131)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[test/map.spec.js, line 114 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-Kr8jfGeB_PO7mGNTMQn:-KrT51U6V6AvU18y9GzD:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/map.spec.js#L114)):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
*[src/helpers/map.js, line 55 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTHV10NlD47HSKF2AD:-KrTHV10NlD47HSKF2AE:b-oh0rfx) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L55)):*
> ```JavaScript
>   const gridTile = init2dArray(CPTL, CPTL, 1);
> 
>   if (!propertyFromId(tileID, 'traversable')) { // tile is not fully traversable
> ```

make this more readable as per our discussion

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
*[src/helpers/map.js, line 67 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTIVR1oIl0K5a1v_AR:-KrTIVR1oIl0K5a1v_AS:bbrjnfp) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L67)):*
> ```JavaScript
>             gridTile[i][j] = 0;
>           } // tile is noncircular and nontraversable
>         } else if (isTileType(tileID, 'ANGLE_WALL_1')) {
> ```

isTileType sounds like a check if the string is a real tile

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
*[test/tiles.spec.js, line 62 at r7](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KqxBaLEoVh7KgXWZ-g9-r7-62:-KrTJ5UEb4vSEKkGh_qs:b2nif2d) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/eaac4a7fd97f3febba19bf72532afd3b05daa199/test/tiles.spec.js#L62)):*
<details><summary><i>Previously, shanek21 (Shane Kelly) wrote…</i></summary><blockquote>

Something like that, I'm still not exactly sure hahaha
</blockquote></details>

wat

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
Reviewed 1 of 6 files at r7, 1 of 3 files at r9.
Review status: 5 of 8 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
Reviewed 1 of 6 files at r7, 3 of 3 files at r11.
Review status: 5 of 8 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
Reviewed 2 of 3 files at r12.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/helpers/map.js, line 55 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTHV10NlD47HSKF2AD:-KrTKTstyZM63uDepbLi:b32w0x) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L55)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

make this more readable as per our discussion
</blockquote></details>

OK

---

*[src/helpers/map.js, line 67 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTIVR1oIl0K5a1v_AR:-KrTKZFKwGEUcnjHuQcN:b32w0x) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L67)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

isTileType sounds like a check if the string is a real tile
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

billmwong commented:
Reviewed 1 of 3 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

davidabrahams commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/helpers/map.js, line 55 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTHV10NlD47HSKF2AD:-KrTJW1PseTxwE20EZJZ:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L55)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

make this more readable as per our discussion
</blockquote></details>

Done.

---

*[src/helpers/map.js, line 67 at r11](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrTIVR1oIl0K5a1v_AR:-KrTKqKBzFV3la40jFdU:b-896fix) ([raw file](https://github.com/chauncy-crib/tagprobot/blob/15b47b0b145a3815e15c3adb88fb9e8ef5ea7161/src/helpers/map.js#L67)):*
<details><summary><i>Previously, billmwong (Bill Wong) wrote…</i></summary><blockquote>

isTileType sounds like a check if the string is a real tile
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

shanek21 commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110#-KrCGokbjzb9SBf67Icr:-KrTMz6nfiofe0gmJWZJ:b32w0x) (no related file):*
<details><summary><i>Previously, davidabrahams wrote…</i></summary><blockquote>

:dango:
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->

shanek21 commented:
Reviewed 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/chauncy-crib/tagprobot/110)*
<!-- Sent from Reviewable.io -->
